smolyay/basis.py

```diff
@@ -0,0 +1,226 @@
+import abc
+import math
+
+import numpy
+
+class BasisFunction(abc.ABC):
+    """Abstract class for basis functions.
+
+    Creates a framework for basis functions that are the building blocks
+    for the terms in the surrogate model. 
+    ``points`` property represents all the unique 1D points of the 
+    basis function associated with a Smolyak index described by the 
+    class IndexGrid.
+
+    """
+
+    def __init__(self):
+        self._points = []
+
+    @abc.abstractproperty
+    def points(self):
+        """The points of the basis function assigned to Smolyak indices"""
+        pass
+
+
+    @abc.abstractmethod
+    def __call__(self,x):
+        """Compute term of basis function
+        Returns the output of the basis function with input x
+
+        Parameters
+        ----------
+        x : float
+            input
+
+        Returns
+        -------
+        y : float
+            output
+        """
+        pass
+
+
+class ChebyshevFirstKind(BasisFunction):
+    r"""Basis function Chebyshev polynomials of the first kind
+    Chebyshev polynomials are a sequence of polynomials described by the
+    following recurrence relation:
+
+    ..math:
+        T_0(x) = 1
+        T_1(x) = x
+        T_{n+1}(x) = 2xT_n(x) - T_{n-1}(x)
+
+    ``n`` is the degree of the polynomial this object describes
+    ``points`` represents the extrema of the polynomial of this degree on
+    the domain [-1,1]
+
+    The extrema of the polynomial are calculated via the following
+    equation:
+
+    ..math:
+        x_{n,j}^* = -\cos\left((\frac{j-1}{n-1}\pi\right), j = 1,...,n
+        n = 2^{exactness-1}+1
```

Comment:
This formula holds for any `n` right? With the new layout, there is no concept of exactness.
```suggestion
        x_i^* = -\cos\left((\frac{i-1}{n-1}\pi\right), i = 1,...,n
```